### DISCUSSION TITLE
how to return exception from validator schema?

### DISCUSSION THREAD

**Discussion by ghost** [fastapi/fastapi#7989] [Questions] [answered]
**i have schema with validator**
```python
@validator('number')
    def validate_phone(cls, v):
        if not validate_mobile_number(v):
            return {'msg': 'incorrect mobile number'}
        return v
```
when i send query with correct or incorrect number, i give response like:
```json
{
  "detail": [
    {
      "loc": [
        "body",
        "user",
        "number"
      ],
      "msg": "expected string or bytes-like object",
      "type": "type_error"
    }
  ]
}
```
but where is my exception?

**Comment by prostomarkeloff** [comment]
Hmm. May be should you add `pre=True` to validator decorator? It means what your validator will be called before default validators. You can see more about it in pydantic docs.

**Comment by ghost** [comment]
@prostomarkeloff i'm tried 'pre' and 'always' but it not working(

**Comment by dmontagu** [comment]
@foozzi 

You aren't using a `pydantic.validator` properly -- you need to raise an error to cause a ValidationError, not return a dict. You probably don't need to set `pre=True` or `always=True` if you are just trying to validate if a string is a valid phone number, but you should only ever return a value that passes your validation. If the input fails validation, you need to raise an error.

If you are trying to get a custom error message, you may need to subclass `pydantic.errors.PydanticValueError`, for example, like this (taken out of `pydantic.errors`):
```python
class StrictBoolError(PydanticValueError):
    msg_template = 'value is not a valid boolean'
```
and then raise that error in your validator.

You should probably also read the pydantic validator docs to make sure you understand what is going on.

If you are trying to change the 
You have to raise a `ValueError`, `TypeError`, or `AssertionError` inside a validator in order to get an error output. You can read the pydantic docs about validators for more details.

**Comment by ghost** [comment]
@dmontagu Thank you! 
but i'm tried this:
```python
class UserBase(BaseModel):
    number: int
    
    @validator('number')
    def validate_phone(cls, v):
        print(validate_mobile_number(v))
        if validate_mobile_number(v) is False:
            raise TypeError(f'test')
        return v
```
and nothing changed...

i'm read docs for pydantic for validators

**Comment by dmontagu** [comment]
Like I said, if you want to change the error message, you have to subclass `PydanticValueError` (or `PydanticTypeError` or similar).

**Comment by prostomarkeloff** [comment, marked as answer]
@foozzi 
For example:

```py
from fastapi import FastAPI
import pydantic

app = FastAPI()


class BadIdError(pydantic.PydanticValueError):
    msg_template = "You sent a wrong id!"


class MyUser(pydantic.BaseModel):
    id: int
    name: str

    @pydantic.validator("id")
    def validate_id(cls, v):
        if v + 1 != 10:
            raise BadIdError()
        return v


@app.post("/create")
async def create_user(_: MyUser):
    return {"status": "ok"}
```

![image](https://user-images.githubusercontent.com/28061158/68535713-e5c2ac00-0357-11ea-8aba-830d88081487.png)

**Comment by tiangolo** [comment]
Thanks a lot for the help @dmontagu  and @prostomarkeloff ! :bowing_man: :clap: 

Does that solve your use case @foozzi ? Wanna close this issue?

**Comment by ghost** [comment]
> Thanks a lot for the help @dmontagu and @prostomarkeloff ! bowing_man clap
> 
> Does that solve your use case @foozzi ? Wanna close this issue?

yep! thx you all

**Comment by sml1991** [comment]
Hey people ! I have a similar problem.  Consider the following (shortened) code:
```
class JobToBeScheduled(BaseModel):
    run_date: datetime
    # i left out the other fields

    @validator("run_date")
    def ensure_date_is_future_date(cls, v):
        bln_tz = pytz.timezone("Europe/Berlin")
        if not v.astimezone(bln_tz) > datetime.now(bln_tz):
            raise ValueError("Must be future date !")
        return v
```
       
It's supposed to run some simple validation on the run_date field. The problem is this: The Swagger Interface displays the ValueError message as expected.

> {
>   "detail": [
>     {
>       "loc": [
>         "body",
>         "run_date"
>       ],
>       "msg": "Must be future date !",
>       "type": "value_error"
>     }
>   ]
> }

But the FastAPI logs just say "422 Unprocessable Entity". How can i get that specific ValueError of mine to be sent as the status of a failed post-request instead of that generic 422 ? @dmontagu  do you have a solution ?
